### DISCUSSION TITLE
Get Bounding Boxes through Protobuf Stack

### DISCUSSION THREAD

**Discussion by youngbrett** [iTwin/itwinjs-core#2464] [Q&A] [answered]
Hi --

I'm building off the Unity example with the Google Protobuf stack.  Is there a code example of fetching only bounding boxes from an iTwin?  I want to delay loading geometry until it is visible and can use bounding boxes with occlusion culling to determine what's hidden.  

Thanks!

**Comment by mgooding** [comment, marked as answer]
The Unity example is determining mesh request order using a sorted list of the element AABBs in the camera's frustum, see here: https://github.com/iTwin/imodel-unity-example/blob/master/unity/Assets/Bentley/Scripts/GraphicsStreaming.cs#L108

You could adapt this example to get the N biggest AABBs (or N biggest in some area, or just a sorted list of all, etc). https://www.itwinjs.org/learning/spatialqueries/ is a helpful resource for this.

If you have a specific request, I can cook up an example on a branch of imodel-unity-example.

**Comment by youngbrett** [comment]
Ah jeez - I missed the sorted list of AABB's.  That's all I need.  Thanks, @mgooding!

**Reply by youngbrett** [reply]
Looked into this a bit more.  The sorted list of AABB's comes back through a spatial query but we don't have any info about the AABB's except their ID and their order in the list.  What I'm looking for is a request / reply that returns the element ID's and AABB attributes as an intermediate step for a mesh request.  Then, you can instantiate a GameObject with a renderer that matches the size and position of the AABB with a Renderer.OnBecameVisible() event, which would then request the actual mesh, texture, properties.  It's kind of a hybrid occlusion pipeline.  

Does that sound like it would be performant?  

My frustrum is usually set to be pretty deep so sorting by volume will pull a lot of meshes.  I guess the spatial query could be rewritten to pull the closest AABB's first, but if you imagine standing back and looking at a building, you really want the roof and exterior walls first.  All this assumes you'd select whether you use this mode, which is kinda focused on design viz.  We're also using this stack to pull all the geometry for pipe stuff, where visibility doesn't matter.

**Reply by mgooding** [reply]
> Does that sound like it would be performant?

It would be fine from the iTwin.js side... I'm not sure from the Unity side but it's an interesting idea to try out.

Regardless of the approach you take, I think an example that gets AABBs is generally useful. I'll put something together and put it on a branch of imodel-unity-example at least.

**Reply by mgooding** [reply]
@youngbrett - take a look here: https://github.com/iTwin/imodel-unity-example/tree/query-aabbs. This will draw the 50 largest BBs in your Scene view on startup. Note that these are AABBs, but are intentionally sorted by their element-aligned volume here.
